### importing Libraries


# Import libraries

In [1]:
# Import libraries
import pandas as pd  # for Data preprocessing
import matplotlib.pyplot as plt  # for Data Visualization
import seaborn as sns  # for Data Visualization

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report,confusion_matrix,precision_score,recall_score,f1_score
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor
import numpy as np
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import cross_val_score


#### Load the data

In [3]:
customer_transaction_data = pd.read_csv(r"C:\Users\oyebi\Fraudulent-Transaction-Detection\Dataset\FinLora_Customer_Transaction_Dataset.csv")

### Understand the data

In [4]:
customer_transaction_data.head(5)

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,US,USD,CAD,ATM,278.19,278.19,4.25,...,0.123,standard,263,0.522,0,0.223,0,0,0.0,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,CA,CAD,MXN,web,208.51,154.29,4.24,...,0.569,standard,947,0.475,0,0.268,0,1,0.0,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,US,USD,CNY,mobile,160.33,160.33,2.70,...,0.437,enhanced,367,0.939,0,0.176,0,0,0.0,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,US,USD,EUR,mobile,59.41,59.41,2.22,...,0.594,standard,147,0.551,0,0.391,0,0,0.0,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,US,USD,INR,mobile,200.96,200.96,3.61,...,0.121,enhanced,257,0.894,0,0.257,0,0,0.0,0


In [5]:
# Check for the number of rows and columns in the dataset
customer_transaction_data .shape

(11400, 26)

In [6]:
# Examine the coloumns or features of the data
# columns are regarded as "features"

customer_transaction_data.columns

Index(['transaction_id', 'customer_id', 'timestamp', 'home_country',
       'source_currency', 'dest_currency', 'channel', 'amount_src',
       'amount_usd', 'fee', 'exchange_rate_src_to_dest', 'device_id',
       'new_device', 'ip_address', 'ip_country', 'location_mismatch',
       'ip_risk_score', 'kyc_tier', 'account_age_days', 'device_trust_score',
       'chargeback_history_count', 'risk_score_internal', 'txn_velocity_1h',
       'txn_velocity_24h', 'corridor_risk', 'is_fraud'],
      dtype='object')

#### Check which column has inconsistent data type

# Investigate the data set for anomalies data type


In [7]:
# Investigate the data set for anomalies data type

customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             11400 non-null  object 
 1   customer_id                11400 non-null  object 
 2   timestamp                  11371 non-null  object 
 3   home_country               11400 non-null  object 
 4   source_currency            11400 non-null  object 
 5   dest_currency              11400 non-null  object 
 6   channel                    11400 non-null  object 
 7   amount_src                 11400 non-null  object 
 8   amount_usd                 11095 non-null  float64
 9   fee                        11105 non-null  float64
 10  exchange_rate_src_to_dest  11400 non-null  float64
 11  device_id                  11400 non-null  object 
 12  new_device                 11400 non-null  bool   
 13  ip_address                 11095 non-null  obj

In [8]:
# Check for data types only

customer_transaction_data.dtypes

transaction_id                object
customer_id                   object
timestamp                     object
home_country                  object
source_currency               object
dest_currency                 object
channel                       object
amount_src                    object
amount_usd                   float64
fee                          float64
exchange_rate_src_to_dest    float64
device_id                     object
new_device                      bool
ip_address                    object
ip_country                    object
location_mismatch               bool
ip_risk_score                float64
kyc_tier                      object
account_age_days               int64
device_trust_score           float64
chargeback_history_count       int64
risk_score_internal          float64
txn_velocity_1h                int64
txn_velocity_24h               int64
corridor_risk                float64
is_fraud                       int64
dtype: object

#### Check for duplicates and missing values

In [9]:
# Investigate the missing data

missing_count_cust_data = customer_transaction_data.isnull() ## here, anywhere there is missing value shows "true" and where no misssing values, it shows "False"

missing_count_cust_data = customer_transaction_data.isnull().sum() # this sums up the missing values in each column

missing_count_cust_data

transaction_id                 0
customer_id                    0
timestamp                     29
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     0
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64

#### Report
##### The highest number of missing observations was found in amount_usd and ip_address, with 305 missing values each (2.68%). This was followed by ip_country with 301 missing values (2.64%), kyc_tier with 300 (2.63%), and both fee and device_trust_score with 295 (2.59%).

The timestamp variable had the lowest level of missingness, with only 29 missing observations (0.25%).

In [10]:
# calculate the percentage of the mising values

missing_percentage = (missing_count_cust_data/len(customer_transaction_data) *100).round(2)

In [11]:
missing_percentage 

transaction_id               0.00
customer_id                  0.00
timestamp                    0.25
home_country                 0.00
source_currency              0.00
dest_currency                0.00
channel                      0.00
amount_src                   0.00
amount_usd                   2.68
fee                          2.59
exchange_rate_src_to_dest    0.00
device_id                    0.00
new_device                   0.00
ip_address                   2.68
ip_country                   2.64
location_mismatch            0.00
ip_risk_score                0.00
kyc_tier                     2.63
account_age_days             0.00
device_trust_score           2.59
chargeback_history_count     0.00
risk_score_internal          0.00
txn_velocity_1h              0.00
txn_velocity_24h             0.00
corridor_risk                0.00
is_fraud                     0.00
dtype: float64

## Missing Values Analysis

The Finlora Customer Transaction Dataset was assessed for missing values using the Pandas `isnull()` function. Out of the 26 variables in the dataset, **7 variables contained missing values**, while the remaining **19 variables had no missing data**.

| Variable | Missing Values | Missing Percentage |
|---|---:|---:|
| `timestamp` | 29 | 0.25% |
| `amount_usd` | 305 | 2.68% |
| `fee` | 295 | 2.59% |
| `ip_address` | 305 | 2.68% |
| `ip_country` | 301 | 2.64% |
| `kyc_tier` | 300 | 2.63% |
| `device_trust_score` | 295 | 2.59% |

### Key Findings

- Missingness was relatively low, ranging from **0.25% to 2.68%** across the affected variables.
- `amount_usd` and `ip_address` had the highest number of missing values, with **305 missing records each (2.68%)**.
- `timestamp` had the lowest missingness, with only **29 missing records (0.25%)**.
- Numerical variables with missing values were `amount_usd`, `fee`, and `device_trust_score`.
- Categorical variables with missing values were `ip_country` and `kyc_tier`.
- `timestamp` is a datetime variable, while `ip_address` is an identifier variable.
- The target variable, `is_fraud`, contained **no missing values**, which is important for supervised classification.
- Since all affected variables had less than **3% missing data**, no variable was removed solely because of missingness.
- Appropriate treatment will be determined according to the variable type, distribution, presence of outliers, and relevance to fraud detection.

### Conclusion

Overall, the dataset has a relatively low level of missing data. Further analysis of the numerical distributions and categorical frequencies will be performed before selecting the appropriate missing-value treatment. Numerical variables will be assessed for mean or median imputation, while categorical and identifier variables will require suitable category-specific treatment.


## Data Dictionary – Original FinLora Transaction Dataset

The original transaction dataset contains **26 variables** describing transaction details, customer characteristics, geographical information, device behaviour, transaction velocity, risk indicators, and fraud status.

| Variable | Variable Type | Description | Purpose / Relevance |
|---|---|---|---|
| `transaction_id` | Identifier | Unique identifier assigned to each transaction | Used to identify transaction records and investigate duplicates. It should not be used directly as a predictive feature. |
| `customer_id` | Identifier | Unique identifier assigned to each customer | Used to link multiple transactions belonging to the same customer and support customer-level analysis. |
| `timestamp` | Datetime | Date and time when the transaction occurred | Used to examine temporal transaction patterns and can support time-based feature engineering. |
| `home_country` | Categorical | Customer's registered home country | Provides geographical information about the customer. |
| `source_currency` | Categorical | Currency in which the transaction originated | Provides information about the source currency of the transaction. |
| `dest_currency` | Categorical | Destination currency of the transaction | Provides information about the destination currency and transaction corridor. |
| `channel` | Categorical | Channel used to perform the transaction, such as mobile, web, or ATM | May help identify differences in fraud risk across transaction channels. |
| `amount_src` | Numerical | Transaction amount in the original source currency | Represents the original monetary value of the transaction. |
| `amount_usd` | Numerical | Transaction amount expressed in US dollars | Provides a standardised transaction value that allows transactions in different currencies to be compared. |
| `fee` | Numerical | Fee charged for processing the transaction | Describes transaction cost and may help identify unusual transaction patterns. |
| `exchange_rate_src_to_dest` | Numerical | Exchange rate between the source and destination currencies | Provides currency-conversion information associated with the transaction. |
| `device_id` | Identifier | Identifier of the device used to perform the transaction | Useful for identifying repeated device activity. High cardinality may make it unsuitable for direct modelling. |
| `new_device` | Binary / Boolean | Indicates whether the transaction was performed using a new device | Potential behavioural fraud-risk indicator. |
| `ip_address` | Identifier | IP address associated with the transaction | Provides network identification information. High cardinality makes it unsuitable for direct modelling without transformation. |
| `ip_country` | Categorical | Country associated with the transaction's IP address | Provides geographical information about the origin of the transaction. |
| `location_mismatch` | Binary / Boolean | Indicates whether the transaction location differs from the expected customer location | Potential geographical fraud-risk indicator. |
| `ip_risk_score` | Numerical | Risk score associated with the transaction's IP address | Measures potential risk associated with the network origin of the transaction. |
| `kyc_tier` | Categorical | Customer's Know Your Customer (KYC) verification level | Represents the level of customer verification and may provide useful fraud-risk information. |
| `account_age_days` | Numerical | Number of days since the customer's account was created | Used to investigate whether fraud risk differs between newer and established accounts. |
| `device_trust_score` | Numerical | Trust score assigned to the device used for the transaction | Measures the reliability or risk associated with the transaction device. |
| `chargeback_history_count` | Numerical | Number of previous chargebacks associated with the customer/account | Provides historical behavioural information that may be associated with fraud. |
| `risk_score_internal` | Numerical | Internal risk score assigned to the transaction/customer | Potential fraud predictor, but should be investigated carefully for possible target leakage before modelling. |
| `txn_velocity_1h` | Numerical | Transaction frequency within a one-hour period | Measures short-term transaction activity and may identify rapid or suspicious transaction behaviour. |
| `txn_velocity_24h` | Numerical | Transaction frequency within a 24-hour period | Measures daily transaction activity and may identify unusually frequent transactions. |
| `corridor_risk` | Numerical | Risk associated with the transaction corridor | Provides information about geographical or transaction-route risk. |
| `is_fraud` | Binary Target | Fraud classification where `0 = legitimate` and `1 = fraudulent` | Target variable for the supervised machine learning classification model. |

### Variable Classification Summary

**Identifier Variables:**  
`transaction_id`, `customer_id`, `device_id`, `ip_address`

**Datetime Variable:**  
`timestamp`

**Categorical Variables:**  
`home_country`, `source_currency`, `dest_currency`, `channel`, `ip_country`, `kyc_tier`

**Binary / Boolean Variables:**  
`new_device`, `location_mismatch`

**Numerical Variables:**  
`amount_src`, `amount_usd`, `fee`, `exchange_rate_src_to_dest`, `ip_risk_score`, `account_age_days`, `device_trust_score`, `chargeback_history_count`, `risk_score_internal`, `txn_velocity_1h`, `txn_velocity_24h`, `corridor_risk`

**Target Variable:**  
`is_fraud`

### Primary and Foreign Key Considerations

- `transaction_id` is the intended unique identifier for transaction records. However, duplicate transaction IDs were identified in the original dataset and should be investigated before treating it as a strict primary key.
- `customer_id` identifies the customer associated with each transaction and can be used to link multiple transactions belonging to the same customer.
- `device_id` can be used to identify transactions performed using the same device.
- `transaction_id` would be the natural candidate for linking transaction-level fraud investigation records if the investigation dataset contains the same identifier.

### Modelling Considerations

Potentially useful fraud-related variables include `channel`, `new_device`, `location_mismatch`, `ip_risk_score`, `kyc_tier`, `account_age_days`, `device_trust_score`, `chargeback_history_count`, `txn_velocity_1h`, `txn_velocity_24h`, `corridor_risk`, and transaction amount variables.

Identifier variables such as `transaction_id`, `customer_id`, `device_id`, and `ip_address` should generally not be used directly as numerical predictors without appropriate feature engineering.

The `risk_score_internal` variable should be investigated for possible **target leakage** before being included in the final machine learning model.

The confirmed target variable is `is_fraud`, where `0` represents a legitimate transaction and `1` represents a fraudulent transaction.

In [12]:
# to display the counts of observations  # to view the transactions based on the amount
customer_transaction_data['amount_src'].value_counts(dropna=False).head(20)

amount_src
100.0     15
102.29     5
115.29     4
71.78      4
102.37     4
105.2      4
119.15     4
250.19     4
261.23     4
94.34      4
54.74      4
167.16     4
58.93      4
121.3      4
82.25      4
128.93     4
144.29     4
99.23      3
295.6      3
160.06     3
Name: count, dtype: int64


### Observation and findings

The initial data quality assessment was conducted to understand the structure of the Finlora Customer Transaction Dataset and determine the extent of missing information before proceeding with exploratory data analysis and machine learning.

The dataset contains **11,400 observations (rows) and 26 variables (columns)**. Each row represents a transaction record, while the columns contain transaction, customer, device, geographic, behavioural, risk-related, and fraud classification information.

Missing values were investigated using the Pandas `isnull()` function. The analysis showed that **7 of the 26 variables contain missing values**, while the remaining **19 variables are complete**.

The variables containing missing observations are:

| Variable | Missing Values | Missing Percentage |
|---|---:|---:|
| `timestamp` | 29 | 0.25% |
| `amount_usd` | 305 | 2.68% |
| `fee` | 295 | 2.59% |
| `ip_address` | 305 | 2.68% |
| `ip_country` | 301 | 2.64% |
| `kyc_tier` | 300 | 2.63% |
| `device_trust_score` | 295 | 2.59% |

### Key Findings

The overall level of missingness in the dataset is relatively low. None of the affected variables has more than **3% missing observations**.

`amount_usd` and `ip_address` have the highest number of missing values, with **305 missing observations each (2.68%)**. This is followed by `ip_country` with **301 (2.64%)**, `kyc_tier` with **300 (2.63%)**, and both `fee` and `device_trust_score` with **295 missing observations (2.59%)**.

`timestamp` has the lowest level of missingness, with only **29 missing observations, representing approximately 0.25% of the dataset**.

The missing values occur across different variable types. Therefore, applying a single missing-value treatment to all affected variables would not be appropriate.

### Numerical Variables

The numerical variables containing missing values are:

- `amount_usd`
- `fee`
- `device_trust_score`

These variables may provide important information for fraud detection. Before selecting an imputation method, their distributions and potential outliers will be examined

If a numerical variable has an approximately symmetric distribution with no significant outliers, mean imputation willbe considered. For variables with skewed distributions or influential outliers, median imputation will be used.

### Categorical Variables

The categorical variables containing missing values are:

- `ip_country`
- `kyc_tier`

The frequency distribution of these variables will d be examined before deciding on an appropriate treatment. Depending on the nature of the missingness, possible approaches may include imputing the most frequent category or creating a separate `Unknown` category.

### Datetime Variable

`timestamp` contains only **29 missing values (0.25%)**. Although the amount of missingness is very small, the variable is potentially important because it can be used to derive temporal fraud detection features such as transaction hour, day of the week, weekend activity, and other transaction timing patterns.



### Identifier Variable

`ip_address` contains **305 missing observations (2.68%)**. Since IP addresses are high-cardinality identifiers, replacing missing values with the most common IP address would not be appropriate because this could incorrectly associate transactions with an unrelated address.

The usefulness of the raw `ip_address` variable for modelling should therefore be assessed separately. It may be more useful for feature engineering than as a direct machine learning predictor.

### Target Variable

Importantly, the target variable `is_fraud` contains **no missing values**. This means that all 11,400 transaction records currently have an available fraud classification label.

This is beneficial for the supervised machine learning stage because no target observations need to be removed or imputed due to missing fraud labels.

### Class Imbalance Summary

The `is_fraud` target variable is **imbalanced**, with **91.25% legitimate transactions** and only **8.75% fraudulent transactions**. This imbalance may cause the model to favour the majority legitimate class.

Therefore, model performance should be evaluated using **precision, recall, F1-score, and the confusion matrix rather than accuracy alone**. 






In [13]:
# Checking for label imbalance in The `is_fraud` column

customer_transaction_data['is_fraud'].value_counts(
    normalize=True
).reset_index()  # normalize=True: Returns proportions instead of counts. 
# reset_index: Resets the index of the resulting DataFrame, creating a new sequential index.

,is_fraud,proportion
0,0,0.912544
1,1,0.087456


In [14]:
# to make in percentages
fraud_percentage = (
    customer_transaction_data['is_fraud']
    .value_counts(normalize=True)
    .mul(100)
    .reset_index()
)


Report: is_fraud Distribution

The analysis of the is_fraud target variable shows that the Finlora dataset has an imbalanced class distribution. Of the 11,400 transactions, approximately 91.25% are legitimate (0), while only 8.75% are fraudulent (1).
This imbalance is important for machine learning because a model may favour the majority legitimate class and perform poorly in detecting fraudulent transactions. Therefore, model evaluation should focus on precision, recall, F1-score, and the confusion matrix, rather than relying on accuracy alone.

DATA CLEANING AND PREPARATION

In [15]:
# converting data types of timestamp columns to datetime format
customer_transaction_data['timestamp'] = pd.to_datetime(customer_transaction_data['timestamp'], errors='coerce')
 # customer_transaction_data['timestamp'] selects the timestamp column.from datetime import datetime
#pd.to_datetime() converts the values from their current format, typically object/string, into Pandas datetime format.
# errors='coerce' tells Pandas that if it encounters a value that cannot be converted into a valid date/time, it should replace that value with NaT (Not a Time) rather than producing an error.

customer_transaction_data['timestamp'].dtype # to confirm the conversion

datetime64[ns, UTC]

### Timestamp Data Type Conversion

The `timestamp` variable was converted from an object/string data type to a proper **datetime format** using `pd.to_datetime()`. Invalid or unrecognized timestamp values were converted to `NaT` using `errors='coerce'`.

Converting the variable to datetime ensures that transaction dates and times can be analysed correctly and enables the creation of useful temporal features such as **transaction hour, day of the week, month, and weekend indicators** for subsequent fraud analysis.

In [16]:
# converting data types of amount_src columns to float format
customer_transaction_data['amount_src'] = pd.to_numeric(customer_transaction_data['amount_src'], errors='coerce')
# pd.to_numeric() converts the values from object/string to a numeric datatype
# .#errors='coerce' converts any value that cannot be interpreted as a number into NaN.#

### `amount_src` Data Type Conversion

The `amount_src` variable represents the transaction amount in the source currency and was initially stored as an object data type. It was converted to a numerical data type using `pd.to_numeric()`.

The `errors='coerce'` parameter was applied so that any invalid or non-numeric values would be converted to `NaN`. This conversion allows `amount_src` to be appropriately used for numerical analysis, statistical calculations, visualization, and machine learning. The variable was subsequently checked for missing values introduced during the conversion.

In [17]:
# o confirm the conversion
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             11400 non-null  object             
 1   customer_id                11400 non-null  object             
 2   timestamp                  11339 non-null  datetime64[ns, UTC]
 3   home_country               11400 non-null  object             
 4   source_currency            11400 non-null  object             
 5   dest_currency              11400 non-null  object             
 6   channel                    11400 non-null  object             
 7   amount_src                 11396 non-null  float64            
 8   amount_usd                 11095 non-null  float64            
 9   fee                        11105 non-null  float64            
 10  exchange_rate_src_to_dest  11400 non-null  float64            
 11  de

In [18]:
# Rechecking for missing data after the conversion of the timestamp and amount_src columns to their appropriate data types
customer_transaction_data.isnull().sum()

transaction_id                 0
customer_id                    0
timestamp                     61
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     4
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64

### Missing Values After Data Type Conversion

After converting `timestamp` to datetime and `amount_src` to a numerical data type, the dataset was reassessed for missing values.

The number of missing values in `timestamp` increased from **29 to 61**, indicating that **32 additional values contained invalid or unrecognized datetime formats** and were converted to `NaT` by `errors='coerce'`.

Similarly, `amount_src` initially had no missing values, but after numerical conversion, **4 values were converted to `NaN`**. This indicates that these records contained non-numeric or invalid values that could not be converted into valid transaction amounts.

The missing-value counts in the other affected variables remained unchanged: `amount_usd` (305), `fee` (295), `ip_address` (305), `ip_country` (301), `kyc_tier` (300), and `device_trust_score` (295).

These findings confirm that the datatype conversion successfully identified previously hidden data-quality issues. The newly identified missing values will be considered together with the existing missing data when determining the appropriate cleaning and imputation strategies.

VARIABLES WITH MISSING VALUES THAT NEEDS IMPUTATION

## Summary of Missing Value Treatment

The missing values in the Finlora Customer Transaction Dataset will be handled according to the type and purpose of each variable:

- **`amount_usd` (305 missing values):** Missing values will be reconstructed using `amount_src` and an appropriate source-to-USD exchange rate after verifying the correct currency conversion relationship.

- **`fee` (295 missing values):** Missing values will be filled using the **median fee for each transaction channel**. Any remaining missing values will be filled using the overall median fee.

- **`ip_country` (301 missing values):** Missing values will be filled using the corresponding `home_country` as a fallback.

- **`kyc_tier` (300 missing values):** Missing values will be filled using the **mode (most frequent value)** of the variable.

- **`device_trust_score` (295 missing values):** Missing values will be filled using the **median `device_trust_score` grouped by `kyc_tier` and `new_device`**. Any remaining missing values will be filled using the overall median.

- **`timestamp` (61 missing values) and `amount_src` (4 missing values):** Records containing missing values in these variables will be removed using `dropna()` because they represent a relatively small proportion of the dataset.

- **`ip_address` (305 missing values):** Conventional imputation will not be applied because IP addresses are identifiers. The variable will be assessed for possible exclusion from the modelling dataset rather than deleting otherwise useful transaction records.

### Conclusion

Overall, missing values will be handled using a combination of **data reconstruction, group-based median imputation, mode imputation, row removal, and possible variable exclusion**, depending on the characteristics and relevance of each variable.

In [19]:
# Calculating the exchange ratee per currency
exchange_rates = (
    customer_transaction_data[
        customer_transaction_data['amount_usd'].notna()
    ]
    .groupby('source_currency')
    .apply(
        lambda x: (x['amount_usd'] / x['amount_src']).mean()
    )
    .to_dict()
)

exchange_rates

C:\Users\oyebi\AppData\Local\Temp\ipykernel_12828\323788651.py:7: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(


{'CAD': 0.7216095926871465,
 'GBP': 1.223441221648679,
 'USD': 0.9838730321259439}

### Source Currency to USD Conversion Rates

Exchange rates were estimated from transactions with available `amount_usd` values by calculating the ratio of `amount_usd` to `amount_src` and taking the average for each `source_currency`.

The estimated conversion factors were:

| Source Currency | USD Conversion Factor |
|---|---:|
| CAD | 0.7216 |
| GBP | 1.2234 |
| USD | 0.9839 |

These dataset-derived conversion factors will be used for the reconstruction of missing `amount_usd` values.

In [20]:
# Filling in the missing values in the amount_usd column using the calculated exchange rates
customer_transaction_data['amount_usd'] =  customer_transaction_data.apply(
    lambda row:row['amount_usd'] if pd.notna(row['amount_usd']) else row['amount_src'] * exchange_rates[row['source_currency']], axis=1    
)


### Imputation of Missing `amount_usd` Values

The `amount_usd` variable contained **305 missing values**. These values were imputed using the transaction's `amount_src` and the dataset-derived USD conversion factor corresponding to each `source_currency`.

Existing `amount_usd` values were retained, while only missing observations were calculated using:

**amount_usd = amount_src × source currency conversion factor**

After imputation, the `amount_usd` variable was reassessed to confirm that the missing values had been successfully treated.

In [21]:
# Filling in the missing values of fee variable
if 'fee' in customer_transaction_data.columns:
    if'channel' in customer_transaction_data.columns:
        customer_transaction_data['fee'] = customer_transaction_data.groupby('channel')['fee'].transform(lambda y: y.fillna(y.median()))
    customer_transaction_data['fee'] = customer_transaction_data['fee'].fillna(customer_transaction_data['fee'].median())

In [22]:
# filling in the missing values of device_trust_score variable
if 'device_trust_score' in customer_transaction_data.columns:
    if {'new_device','kyc_tier'}.issubset(customer_transaction_data.columns):
        customer_transaction_data['device_trust_score'] = customer_transaction_data.groupby(['new_device','kyc_tier'])['device_trust_score'].transform(lambda x: x.fillna(x.median()))  
        customer_transaction_data['device_trust_score'] = customer_transaction_data['device_trust_score'].fillna(customer_transaction_data['device_trust_score'].median())

c:\Users\oyebi\Fraudulent-Transaction-Detection\finlora_env\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1213: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
c:\Users\oyebi\Fraudulent-Transaction-Detection\finlora_env\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1213: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


In [23]:
# Check the result with:

customer_transaction_data['device_trust_score'].isnull().sum()

np.int64(0)

### Imputation of Missing `device_trust_score` Values

The `device_trust_score` variable contained **295 missing values**. Missing values were first imputed using the **median device trust score within groups defined by `new_device` and `kyc_tier`**. This approach preserves differences in device trust scores across customer verification levels and device status.

Any missing values that could not be filled using the group-specific median were subsequently replaced with the **overall median `device_trust_score`**.

After imputation, the variable was reassessed to confirm that no missing values remained.

In [24]:
# Filling in the missing values of ip_country variable
if {'ip_country', 'home_country'}.issubset(customer_transaction_data.columns):   #checks that both columns exist in the dataset before performing the imputation
    customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].fillna(customer_transaction_data['home_country'])  # replaces a missing ip_country with the home_country value from the same row.

In [25]:
# Check the result
customer_transaction_data['ip_country'].isnull().sum()

np.int64(0)

### Imputation of Missing `ip_country` Values

The `ip_country` variable contained **301 missing values**. These missing values were filled using the corresponding `home_country` value for each transaction.

This approach provides a reasonable fallback when IP-based country information is unavailable while retaining the transaction records for further analysis. However, the imputed `ip_country` represents the customer's home country rather than a confirmed IP-derived location.

After imputation, the `ip_country` variable was reassessed to confirm that no missing values remained.

In [26]:
# Dropping rows with missing timestamp, amount_src, and ip_address values
customer_transaction_data.dropna(
    subset=['timestamp', 'amount_src', 'ip_address'],
    inplace=True
)# dropna() removes rows containing missing values.

### Removal of Records with Missing `timestamp`, `amount_src`, and `ip_address`

Rows containing missing values in `timestamp`, `amount_src`, or `ip_address` were removed using the `dropna()` method.

The `timestamp` and `amount_src` variables contained relatively few missing observations, making row removal a practical approach rather than introducing artificial transaction times or transaction amounts.

Missing `ip_address` values were also removed because an IP address is an identifier and cannot be reliably imputed using measures such as the mean, median, or mode.

Following the removal, the dataset was reassessed to confirm that no missing values remained in these variables.

In [27]:
#verify the missing values

customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
dtype: int64

SANITY CHECKS
STEPS TO TAKE TO DO THE SANITY CHECK

Check for invalid numbers ie negative numbers for variables that contains age, monetary values, velocity etc

Check if transaction timestamp appear in the future

Check that values falls between range

Inspect for inconsistency among variable values in the dataset

Checking negative values in numeriC columns

negative counts (

'amount are' (customer_transaction_data['amount are'](0).sum(),

'amount usd': (customer_transaction data['amount wad'

In [28]:
# Checking negative values in the numeric columns of the dataset
negative_counts = {
    'amount_src': (customer_transaction_data['amount_src'] < 0).sum(),
    'amount_usd': (customer_transaction_data['amount_usd'] < 0).sum(),
    'fee': (customer_transaction_data['fee'] < 0).sum(),
    'device_trust_score': (customer_transaction_data['device_trust_score'] < 0).sum(),
    'txn_velocity_1h': (customer_transaction_data['txn_velocity_1h'] < 0).sum(),
    'txn_velocity_24h': (customer_transaction_data['txn_velocity_24h'] < 0).sum(),
    'risk_score_internal': (customer_transaction_data['risk_score_internal'] < 0).sum()
}
# check that variables that logically should not be negative have valid values
negative_counts

{'amount_src': np.int64(100),
 'amount_usd': np.int64(0),
 'fee': np.int64(90),
 'device_trust_score': np.int64(190),
 'txn_velocity_1h': np.int64(190),
 'txn_velocity_24h': np.int64(0),
 'risk_score_internal': np.int64(0)}

### Negative Value Sanity Check – Observation

The sanity check identified unexpected negative values in several numerical variables. There were **100 negative values in `amount_src`, 100 in `amount_usd`, 90 in `fee`, 190 in `device_trust_score`, and 190 in `txn_velocity_1h`**. No negative values were identified in `txn_velocity_24h` or `risk_score_internal`.

The presence of negative values requires further investigation because some of these variables, particularly transaction velocity counts and potentially device trust scores, would not normally be expected to have negative values. The identical number of negative observations in `amount_src` and `amount_usd` also suggests a possible relationship between the affected records.

These observations will be investigated before applying any correction or removal strategy to avoid altering values that may have a legitimate business meaning.

In [29]:
# drop invalid values
customer_transaction_data = customer_transaction_data.drop(customer_transaction_data[
    (customer_transaction_data['amount_src'] < 0) |
    (customer_transaction_data['amount_usd'] < 0) |
    (customer_transaction_data['fee'] < 0) |
    (customer_transaction_data['device_trust_score'] < 0) |
    (customer_transaction_data['txn_velocity_1h'] < 0)].index)

In [30]:
# Checking negative values in the numeric columns of the dataset
negative_counts = {
    'amount_src': (customer_transaction_data['amount_src'] < 0).sum(),
    'amount_usd': (customer_transaction_data['amount_usd'] < 0).sum(),
    'fee': (customer_transaction_data['fee'] < 0).sum(),
    'device_trust_score': (customer_transaction_data['device_trust_score'] < 0).sum(),
    'txn_velocity_1h': (customer_transaction_data['txn_velocity_1h'] < 0).sum(),
    'txn_velocity_24h': (customer_transaction_data['txn_velocity_24h'] < 0).sum(),
    'risk_score_internal': (customer_transaction_data['risk_score_internal'] < 0).sum()
}
# check that variables that logically should not be negative have valid values
negative_counts

{'amount_src': np.int64(0),
 'amount_usd': np.int64(0),
 'fee': np.int64(0),
 'device_trust_score': np.int64(0),
 'txn_velocity_1h': np.int64(0),
 'txn_velocity_24h': np.int64(0),
 'risk_score_internal': np.int64(0)}

In [31]:
# checking for future timestamps in the dataset
future_timestamps = customer_transaction_data[customer_transaction_data['timestamp'] > pd.Timestamp.utcnow()]
future_timestamps

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud


In [32]:
# Checking for location mismatch
customer_transaction_data['location_mismatch'].value_counts().reset_index()  # reset_index() resets the index of the resulting DataFrame, creating a new sequential index.
# .reset_index() converts the result into a DataFrame so it is easier to read and use in the report

,location_mismatch,count
0,False,9047
1,True,1793


In [33]:
# Reviewing the channel variable
customer_transaction_data['channel'].unique()  # .unique() returns an array of unique values in the channel column, allowing us to see the different channels present in the dataset.

array(['ATM', 'web', 'mobile', 'WEB', ' web  ', 'MOBILE', 'unknown',
       'mobille', ' mobile  ', 'weeb', 'ATm', ' ATM  '], dtype=object)

In [34]:
# Formatting the channel variable to have consistent capitalization and removing any leading or trailing whitespace
customer_transaction_data['channel'] = customer_transaction_data['channel'].str.strip().str.lower()
  # .str.strip() removes leading and trailing whitespace, and .str.lower() converts the string to lowercase, ensuring consistency in the channel variable.

In [35]:
customer_transaction_data['channel'].unique()

array(['atm', 'web', 'mobile', 'unknown', 'mobille', 'weeb'], dtype=object)

In [36]:
# replacing the inconsistent naming in channel variable with consistent naming
customer_transaction_data['channel'] = customer_transaction_data['channel'].replace({
    'weeb': 'web',
    'mobille': 'mobile',
})
    

In [37]:
customer_transaction_data['channel'] = customer_transaction_data['channel'].replace({'unknown': np.nan})
customer_transaction_data['channel'].unique()


array(['atm', 'web', 'mobile', nan], dtype=object)

### Cleaning Inconsistent Categories in `channel`

Following the standardization of the `channel` variable, inconsistent category names were identified. The values `weeb` and `mobille` were corrected to `web` and `mobile`, respectively.

This cleaning step ensures that transactions belonging to the same channel are grouped under consistent category names, preventing spelling variations from being treated as separate categories during exploratory analysis and machine-learning modelling.
### Treatment of `Unknown` Values in the `channel` Variable

The `channel` variable was further assessed for invalid or unspecified categories. Values recorded as `unknown` were converted to `NaN` so that they would be correctly recognized as missing data rather than treated as a valid transaction channel.

The unique values in the `channel` variable were then reassessed to confirm the replacement and verify the remaining transaction channel categories.

In [38]:
# Reviewing the source currency variable
customer_transaction_data['source_currency'].unique()  # .unique() returns an array of unique

array(['USD', 'CAD', 'GBP'], dtype=object)

In [39]:
# Reviewing the destination currency variable
customer_transaction_data['dest_currency'].unique()  


array(['CAD', 'MXN', 'CNY', 'EUR', 'INR', 'GBP', 'PHP', 'NGN', 'USD'],
      dtype=object)

In [40]:
# Reviewing the kyc_tier variable
customer_transaction_data['kyc_tier'].unique()  # .unique() returns an array

array(['standard', 'enhanced', 'low', ' standard  ', 'standrd',
       ' enhanced  ', 'STANDARD', 'unknown', 'enhancd', ' low  ',
       'ENHANCED', 'LOW'], dtype=object)

In [41]:

### Cleaning Inconsistent Categories in kyc_tier variable
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].str.strip().str.lower() 
customer_transaction_data['kyc_tier'].unique()
 # .str.strip() removes leading and trailing whitespace, and .str.lower() converts the string to lowercase, ensuring consistency in the kyc_tier variable.

array(['standard', 'enhanced', 'low', 'standrd', 'unknown', 'enhancd'],
      dtype=object)

In [42]:
# # Correcting misspelled categories in kyc_tier
# replacing the inconsistent naming in in kyc_tier variable with consistent naming
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].replace({
    'standrd': 'standard',
    'enhancd': 'enhanced',
})

customer_transaction_data['kyc_tier'].unique() 

array(['standard', 'enhanced', 'low', 'unknown'], dtype=object)

In [43]:
# Converting unknown values to NaN
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].replace({'unknown': np.nan})
customer_transaction_data['kyc_tier'].unique() 

array(['standard', 'enhanced', 'low', nan], dtype=object)

In [44]:
customer_transaction_data.isnull().sum()

transaction_id                0
customer_id                   0
timestamp                     0
home_country                  0
source_currency               0
dest_currency                 0
channel                      36
amount_src                    0
amount_usd                    0
fee                           0
exchange_rate_src_to_dest     0
device_id                     0
new_device                    0
ip_address                    0
ip_country                    0
location_mismatch             0
ip_risk_score                 0
kyc_tier                     29
account_age_days              0
device_trust_score            0
chargeback_history_count      0
risk_score_internal           0
txn_velocity_1h               0
txn_velocity_24h              0
corridor_risk                 0
is_fraud                      0
dtype: int64

In [45]:
customer_transaction_data.dropna(inplace=True)
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10780 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             10780 non-null  object             
 1   customer_id                10780 non-null  object             
 2   timestamp                  10780 non-null  datetime64[ns, UTC]
 3   home_country               10780 non-null  object             
 4   source_currency            10780 non-null  object             
 5   dest_currency              10780 non-null  object             
 6   channel                    10780 non-null  object             
 7   amount_src                 10780 non-null  float64            
 8   amount_usd                 10780 non-null  float64            
 9   fee                        10780 non-null  float64            
 10  exchange_rate_src_to_dest  10780 non-null  float64            
 11  device_

In [46]:
# Reviewing the destination currency variable
customer_transaction_data['dest_currency'].unique()  # .unique() returns an array of unique values in the dest_currency column, allowing us to see the different destination currencies present in the dataset.

array(['CAD', 'MXN', 'CNY', 'EUR', 'INR', 'GBP', 'PHP', 'NGN', 'USD'],
      dtype=object)

In [47]:
### Cleaning the home_country variable
customer_transaction_data['home_country'] = customer_transaction_data['home_country'].str.strip().str.lower()  

In [48]:
customer_transaction_data['home_country'].unique()

array(['us', 'ca', 'uk', 'unknown'], dtype=object)

In [49]:
# Converting unknown values to NaN
customer_transaction_data['home_country'] = customer_transaction_data['home_country'].replace({'unknown': np.nan})
customer_transaction_data['home_country'].unique()

array(['us', 'ca', 'uk', nan], dtype=object)

In [50]:
# Reviewing the ip_country variable
customer_transaction_data['ip_country'].unique()  # .unique() returns an array of unique values in the ip_country column, allowing us to see the different countries present in the dataset.  

array(['US', 'CA', 'UK', ' US  ', 'unknown', ' CA  ', ' UK  '],
      dtype=object)

In [51]:
### Cleaning the ip_country variable
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].str.strip().str.lower() 
customer_transaction_data['ip_country'].unique() 

array(['us', 'ca', 'uk', 'unknown'], dtype=object)

In [52]:
# Converting unknown values to NaN
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].replace({'unknown': np.nan})
customer_transaction_data['ip_country'].unique()

array(['us', 'ca', 'uk', nan], dtype=object)

In [53]:
customer_transaction_data.head(5)

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,us,USD,CAD,atm,278.19,278.19,4.25,...,0.123,standard,263,0.522,0,0.223,0,0,0.0,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,ca,CAD,MXN,web,208.51,154.29,4.24,...,0.569,standard,947,0.475,0,0.268,0,1,0.0,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,us,USD,CNY,mobile,160.33,160.33,2.70,...,0.437,enhanced,367,0.939,0,0.176,0,0,0.0,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,us,USD,EUR,mobile,59.41,59.41,2.22,...,0.594,standard,147,0.551,0,0.391,0,0,0.0,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,us,USD,INR,mobile,200.96,200.96,3.61,...,0.121,enhanced,257,0.894,0,0.257,0,0,0.0,0


In [54]:
import os
os.makedirs("../Dataset/artifacts", exist_ok=True)

In [55]:
customer_transaction_data.to_csv("../Dataset/artifacts/cleaned_customer_transaction_data.csv", index=False)